src/goto-programs/goto_convert_side_effect.cpp

```diff
@@ -128,8 +128,17 @@ void goto_convertt::remove_assignment(
   // revert assignment in the expression to its LHS
   if(result_is_used)
   {
-    exprt lhs;
-    lhs.swap(to_binary_expr(expr).op0());
+    exprt lhs = to_binary_expr(expr).op0();
+    // assign_* statements can have an lhs operand with a different type than
+    // that of the overall expression, because of integer promotion (which may
+    // have introduced casts to the lhs).
+    if(expr.type() != lhs.type())
+    {
+      // Skip over those type casts, but also
+      // make sure the resulting expression has the same type as before.
+      PRECONDITION(to_typecast_expr(lhs).op().type() == expr.type());
```

Comment:
I would add a DATA_INVARIANT here to check that the typecast is there, and turn that PRECONDITION into a DATA_INVARIANT.

Reply:
Done.